### ISSUE TITLE
Unable to reproduce the α and T relationship

### ISSUE THREAD

**Issue by kainat-altaf-cau** [HelloKevin07/AgileNN#2]
Hello,

Thank you for your contribution in the field. It is quite helpful and interesting. 

I was trying to run the given code. The paper mentions that the network would be biased towards the remote NN because of different complexities between local NN and remote NN. To avoid this biasness, the paper intoduces a soft constraint where changing T to specific values avoid such biasness, Figure 8(b). However, I am unable to reproduce the relationship between T and α.

Given the rest of the code as provided, I tried to use different Ts and get the corresponding **α**s.  Here are my results from the experiments.

- if T = 4,  then α =  0.059

- if T = 5,  then α =  0.048

- if T = 6,  then α =  0.038

- if T = 7,   then α =  0.035

- if T = 8,  then α =  0.020

which means even with changing the T, it is biased towards the remote NN.

Is there anything else which I need to change or consider for reproducing the results mentioned in paper? It would be really helpful if you can guide me to navigate through this issue.

Thanks.

**Comment by HelloKevin07** [comment]
Thank you for your interest and question!

This code base is a cleaned version for end-to-end evaluation but is not for reproducing any preliminary results (i.e. in technical sections) in the paper. To reproduce Figure 8(b), besides T, you will also need an extra loss term to suppress this bias, which is not mentioned in our paper due to the page limit. That loss is defined as `c * (w[:, 0] - w[:, 1])^2`, where `w` is defined [here](https://github.com/HelloKevin07/AgileNN/blob/77c8c9b8e4c8c637a1f889896feea565d6a38e34/models.py#LL76C1-L76C1), and `c` is a hyperparameter that can be tuned between 1e-3 and 1e-5.

I'm not sure how you currently configure your training (e.g., rho and split ratio). If you simply want to bypass this bias issue, you can manually set a static alpha value, say alpha=0.5, without trainability, as we mentioned in the paper. This is also what we did in most of our evaluations.

I guess you may be interested in exploring the learning dynamics of such bias when the remote and local NNs are jointly trained. In my hindsight regarding this paper, the alpha may not be able to reflect the biasness, because any NN weights in previous layers could be correspondingly increased/decreased during training to cancel out the value scaling caused by alpha.

**Comment by kainat-altaf-cau** [comment]
Thank you for your response. 

Can you help me point towards `c * (w[:, 0] - w[:, 1])^2` loss term defined in the provided code? 
I am using the default values where rho = 0.8 and lambda =0.2. Looking into [this](https://github.com/HelloKevin07/AgileNN/blob/77c8c9b8e4c8c637a1f889896feea565d6a38e34/models.py#L87) line, it shows a similar depiction of alpha mentioned in paper, also shown below, which means `w1 = alpha` and `w2 = 1-alpha`.
![Screenshot 2023-05-10 at 12 28 32 PM](https://github.com/HelloKevin07/AgileNN/assets/131257718/ce6d2030-4997-4f5e-b45c-a361f98caa2c)

Maybe I am missing something to find the alpha as it is not defined in the code. My goal is to see how much importance we are actually giving to the local and remoteNN depending on the features set they are provided. I hope it helps to answer my question.

**Comment by HelloKevin07** [comment]
Your understanding is conceptually correct. But in our actual implementation, we took **unnormalized** `w1` and `w2`. That literally means `c * (self.w[:, 0] - self.w[:, 1])^2`,  where [self.w](https://github.com/HelloKevin07/AgileNN/blob/77c8c9b8e4c8c637a1f889896feea565d6a38e34/models.py#LL79C9-L81C10) is the trainable weight of the reweighting layer.

If you want to see how much importance is actually distributed to local and remote for each data sample, you can directly retrieve [local and remote importance](https://github.com/HelloKevin07/AgileNN/blob/77c8c9b8e4c8c637a1f889896feea565d6a38e34/train_agilenn.py#LL268C13-L269C81) from the training loop. Please note that you cannot directly print them out because these variables are not eager tensors (assuming you are familiar with tensorflow). Instead, you can first return them from step(), i.e., `local_imp, remote_imp = step(...)`, and then print them out at runtime.
